diofant/functions/elementary/exponential.py

```diff
@@ -147,7 +147,37 @@ def exp(arg, **kwargs):
 
     diofant.functions.elementary.exponential.log
     """
-    return Pow(S.Exp1, arg, **kwargs)
+
+    def __new__(cls, arg, *opt_args, **kwargs):
```

Comment:
What about `arg1, arg2=None`? If arg2 is given, then arg1 has to be Exp1. Otherwise arg1 becomes arg2 and arg1 is Exp1.
```
if arg2 is not None:
    assert arg1 == Exp1
else:
    arg2 = arg1
    arg1 = Exp1
```

Reply:
This is still the same hack, isn't?

What I actually would like to see - strictly 1-arg function exp (and similar for sqrt and cbrt, btw).

Reply:
Well, it might be easier with sqrt as the argument has the same position as in Pow. I believe that position should be preserved, otherwise some inherited methods relying on the position of arguments may cause problems. How will the pattern marcher work on exp inheriting Pow?

If you really want to define such functions, I would consider supporting something similar to Haskell-like currying.

Reply:
> otherwise some inherited methods relying on the position of arguments

Yes, the major reason for implemented hack with arguments - some Pow's methods use "subclassing-friendly" (tm) way of rebuilding expression, i.e. ``self.func(*self.args)``.

> How will the pattern marcher work on exp inheriting Pow?

What you mean?

```
In [1]:In [1]: x_, y_ = symbols('x,y', cls=Wild)

In [2]: exp(x).match(y_**x_)
Out[2]: {x_: x, y_: ℯ}
```

Reply:
I believe we need a way of creating a new function by to reducing the number of arguments in a function. Python would kind of allow this by defining `exp = lambda x: Pow(Exp1, x)`. This creates a lambda, but the ideal is to have something similar that creates a Function object by partly defining some arguments.

Or maybe allowing Pow to support a single argument call: Pow(Exp1) would create exp, so that Pow(Exp1)(x) == Pow(Exp1, x).

Reply:
> I believe we need a way of creating a new function by to reducing the number of arguments in a function.

And how this will fix Pow's methods, that use "subclassing-frienldy" (tm) way of re-creating expression, namely ``self.func(b, e)``?

> Pow(Exp1) would create exp

Functions are not "first class citizens" in diofant (or sympy), e.g. you can't do ``Add(Function('f'), Function('g'))``.  So, doesn't make sense for me.

Reply:
> Functions are not "first class citizens" in diofant (or sympy), e.g. you can't do Add(Function('f'), Function('g')). So, doesn't make sense for me.

What's your stance on that? Do you support making functions first-class citizens? There's been a long issue about that in SymPy.

Reply:
I'm just saying that mentioned your suggestion doesn't make sense without this.